File: app/src/main/java/com/example/mediaplayer/chosenSong/ChosenSongFragment.java
package com.example.mediaplayer.chosenSong;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Build;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import android.os.IBinder;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.mediaplayer.R;
import com.example.mediaplayer.chosenSong.ChosenSongService.SongBinder;
import com.example.mediaplayer.databinding.FragmentChosenSongBinding;
import com.example.mediaplayer.playlist.PlayListModel;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.Player.EventListener;
import java.util.ArrayList;
import java.util.Objects;
import static com.example.mediaplayer.constants.ACTION_FOREGROUND;
import static com.example.mediaplayer.constants.ACTION_PLAYING_AUDIO_FOREGROUND;
import static com.example.mediaplayer.constants.AUDIO_FOREGROUND_NOTIFICATION;
import static com.example.mediaplayer.constants.CHOSEN_SONG_INDEX;
import static com.example.mediaplayer.constants.LIST_SONG;

/**
 * A simple {@link Fragment} subclass.
 */
public class ChosenSongFragment extends Fragment {
    private ChosenSongViewModel viewModel;
    private FragmentChosenSongBinding binding;
    private ChosenSongService mService;
    private Intent foregroundIntent;
    private String purposeOfFragment;

    public ChosenSongFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentChosenSongBinding.inflate(inflater);
        setUpViewModel();
        startForeground();
        setUpPlayerView();
        return binding.getRoot();
    }



    private void setUpViewModel() {
        ArrayList<PlayListModel> playListModels = null;
        int chosenSongIndex = 0;
        if (getArguments() != null) {
            playListModels = getArguments().getParcelableArrayList(LIST_SONG);
            chosenSongIndex = getArguments().getInt(CHOSEN_SONG_INDEX, 0);
            purposeOfFragment=getArguments().getString(AUDIO_FOREGROUND_NOTIFICATION,ACTION_PLAYING_AUDIO_FOREGROUND);
            //clearing the bundle so when fragment stops no huge parcelable error occurs
            //avoid causing TransactionTooLargeException
            getArguments().clear();
        }
        ChosenSongViewModelFactory factory = new ChosenSongViewModelFactory(playListModels,chosenSongIndex);
        viewModel = ViewModelProviders.of(this, factory).get(ChosenSongViewModel.class);

    }

    private void setUpPlayerView() {
        binding.playerView.setUseController(true);
        binding.playerView.showController();
        binding.playerView.setControllerAutoShow(true);
    }

    private void setSongName(int index) {
        ((TextView) binding.playerView.findViewById(R.id.songName)).setText(viewModel.getPlayListModels().get(index).getTitle());

    }

    private void setActorName(int index) {
        ((TextView) binding.playerView.findViewById(R.id.songActor)).setText(viewModel.getPlayListModels().get(index).getActor());

    }

    private void initializePlayer() {
        ExoPlayer player = mService.getPlayer();
        //Attaching the player to a view
        binding.playerView.setPlayer(player);
        mService.getPlayer().addListener(new EventListener() {
            @Override
            public void onSeekProcessed() {
                //if any changes happened to player transfer these changes to viewmodel Live data
                viewModel.setChosenSongIndex(mService.getPlayer().getCurrentWindowIndex());

            }

        });


    }


    @Override
    public void onStart() {
        super.onStart();
        //binding this fragment to service
        Objects.requireNonNull(getActivity()).bindService(foregroundIntent, connection, Context.BIND_AUTO_CREATE);


    }

    @Override
    public void onStop() {
        super.onStop();
        //unBinded fragment from service
        Objects.requireNonNull(getActivity()).unbindService(connection);

    }


    private void startForeground() {
        foregroundIntent = new Intent(getActivity(), ChosenSongService.class);
        foregroundIntent.setAction(ACTION_FOREGROUND);
        foregroundIntent.putExtra(CHOSEN_SONG_INDEX, viewModel.getChosenSongIndex().getValue());
        foregroundIntent.putParcelableArrayListExtra(LIST_SONG, viewModel.getPlayListModels());
        //Start service:
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Objects.requireNonNull(getActivity()).startForegroundService(foregroundIntent);

        } else {
            Objects.requireNonNull(getActivity()).startService(foregroundIntent);

        }
    }


    /**
     * Defines callbacks for service binding, passed to bindService()
     */
    private ServiceConnection connection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName className,
                                       IBinder service) {
            // We've bound to LocalService, cast the IBinder and get LocalService instance
            SongBinder binder = (SongBinder) service;
            mService = binder.getService();
            initializePlayer();
            viewModel.getChosenSongIndex().observe(getViewLifecycleOwner(), chosenSongIndex -> {
                if(chosenSongIndex!=null) {
                    //to avoid setup player again when configuration changes happen or if player fragment has opened through service from notification
                    //if player is already setuped no need to setup it again
                    if (!viewModel.isServiceCreatedBefore() && !purposeOfFragment.equals(AUDIO_FOREGROUND_NOTIFICATION))
                        mService.setUpPlayer(viewModel.getPlayListModels(), chosenSongIndex);
                    setSongName(chosenSongIndex);
                    setActorName(chosenSongIndex);
                }
            });
            //if this fragment is opened through service then the source of truth is service and set the value of chosenSongIndex from service
            if(purposeOfFragment.equals(AUDIO_FOREGROUND_NOTIFICATION))
            {
                viewModel.setChosenSongIndex(mService.getPlayer().getCurrentWindowIndex());
            }
            //set this service as created before so we do not setupPlayer again
            viewModel.setServiceCreatedBefore();

        }

        @Override
        public void onServiceDisconnected(ComponentName arg0) {

        }
    };

}

File: app/src/main/java/com/example/mediaplayer/chosenSong/ChosenSongFragment.kt
package com.example.mediaplayer.chosenSong

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.example.mediaplayer.chosenSong.ChosenSongService.SongBinder
import com.example.mediaplayer.constants.*
import com.example.mediaplayer.databinding.FragmentChosenSongBinding
import com.example.mediaplayer.playlist.PlayListModel
import com.google.android.exoplayer2.Player.EventListener
import com.google.android.exoplayer2.ui.PlayerView
import kotlinx.android.synthetic.main.exo_player_view.view.*
import java.util.*


data class MediaInfo(var playListModels: ArrayList<PlayListModel>? = null,
                     var chosenSongIndex: Int = 0, val purposeOfFragment: String = ACTION_PLAYING_AUDIO_FOREGROUND)

/**
 * A simple [Fragment] subclass.
 */
class ChosenSongFragment : Fragment() {

    private lateinit var viewModel: ChosenSongViewModel
    private lateinit var binding: FragmentChosenSongBinding
    private lateinit var foregroundIntent: Intent


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        binding = FragmentChosenSongBinding.inflate(inflater)
        setUpViewModel()
        startForeground()
        return binding.root
    }

    private fun getMediaInfo(): MediaInfo {
        if (arguments != null) {
            val playListModels: ArrayList<PlayListModel>? = arguments!!.getParcelableArrayList(LIST_SONG)
            val chosenSongIndex = arguments!!.getInt(CHOSEN_SONG_INDEX, 0)
            val purposeOfFragment = arguments!!.getString(AUDIO_FOREGROUND_NOTIFICATION, ACTION_PLAYING_AUDIO_FOREGROUND)
            //clearing the bundle so when fragment stops no huge parcelable error occurs
            //avoid causing TransactionTooLargeException
            arguments!!.clear()
            return MediaInfo(playListModels, chosenSongIndex, purposeOfFragment)
        }
        return MediaInfo()
    }

    private fun setUpViewModel() {
        val info = getMediaInfo()
        val factory = ChosenSongViewModelFactory(info.playListModels, info.chosenSongIndex, info.purposeOfFragment)
        viewModel = ViewModelProviders.of(this, factory).get(ChosenSongViewModel::class.java)

        //observe if the service is yet initialized or not  so we can setup player and synchronize it with ui
        viewModel.chosenSongService.observe(viewLifecycleOwner, Observer { service ->
            service?.let {
                binding.playerView.initializePlayer(service)
                //if this fragment is opened through service then the source of truth is service and set the value of chosenSongIndex from service
                if (viewModel.purposeOfFragment == AUDIO_FOREGROUND_NOTIFICATION) {
                    viewModel.setChosenSongIndex(service.player.currentWindowIndex)
                }
                //observe if the the the song was changed and based on that we reflect that change on ui
                viewModel.chosenSongIndex.observe(viewLifecycleOwner, Observer { index ->
                    index.let {
                        //to avoid setup player again when configuration changes happen or if player fragment has opened  from notification
                        if (!viewModel.isServiceCreatedBefore && viewModel.purposeOfFragment != AUDIO_FOREGROUND_NOTIFICATION)
                            viewModel.playListModels?.let {
                                service.setUpPlayer(it, index)
                            }
                        //whenever the chosen song index change we update the song info to reflect the current index of song
                        binding.playerView.songDetails(index)
                    }
                })
                //set this service as created before so we do not setupPlayer again
                viewModel.isServiceCreatedBefore = true

            }

        })
    }


    private fun PlayerView.songDetails(index: Int) {
        this.songName.text = viewModel.playListModels!![index].title
        this.songActor.text = viewModel.playListModels!![index].actor


    }


    private fun PlayerView.initializePlayer(service: ChosenSongService) {
        val player = service.player
        //Attaching the player to a view
        this.player = player
        this.useController = true
        this.showController()
        this.controllerAutoShow = true
        service.player.addListener(object : EventListener {

            override fun onSeekProcessed() {
                viewModel.setChosenSongIndex(service.player.currentWindowIndex)
            }

        })


    }


    override fun onStart() {
        super.onStart()
        //binding this fragment to service
        Objects.requireNonNull<FragmentActivity>(activity).bindService(foregroundIntent, connection, Context.BIND_AUTO_CREATE)


    }

    override fun onStop() {
        super.onStop()
        //un Bind fragment from service
        Objects.requireNonNull<FragmentActivity>(activity).unbindService(connection)

    }


    private fun startForeground() {
        foregroundIntent = Intent(activity, ChosenSongService::class.java)
        foregroundIntent.action = ACTION_FOREGROUND
        foregroundIntent.putExtra(CHOSEN_SONG_INDEX, viewModel.chosenSongIndex.value)
        foregroundIntent.putParcelableArrayListExtra(LIST_SONG, viewModel.playListModels)
        //Start service:
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Objects.requireNonNull<FragmentActivity>(activity).startForegroundService(foregroundIntent)

        } else {
            Objects.requireNonNull<FragmentActivity>(activity).startService(foregroundIntent)

        }
    }

    /**
     * Defines callbacks for service binding, passed to bindService()
     */
    private val connection = object : ServiceConnection {

        override fun onServiceConnected(className: ComponentName,
                                        service: IBinder) {
            // We've bound to LocalService, cast the IBinder and get LocalService instance
            val binder = service as SongBinder
            val mService = binder.service
            viewModel.setChosenSongService(mService)

        }

        override fun onServiceDisconnected(arg0: ComponentName) {

        }
    }

}// Required empty public constructor
